src/hardware/sblaster.cpp

```diff
@@ -886,22 +887,53 @@ static std::array<uint8_t, 2> decode_adpcm_4bit(const uint8_t data)
 	        decode_adpcm_portion(data & 0xf, AdjustMap, ScaleMap, LastIndex)};
 }
 
+// Returns a vector of AudioFrames from the source samples. If the Sound Blaster
+// is still warming up then the frames will be silent.
 template <typename T>
-static const T* maybe_silence(const uint32_t num_samples, const T* buffer)
+static std::vector<AudioFrame>& maybe_silence(const T* samples,
+                                              const uint32_t num_samples,
+                                              const FrameType frame_type)
 {
-	if (sb.dsp.warmup_remaining_ms <= 0) {
-		return buffer;
-	}
+	assert(samples && num_samples > 0);
 
-	static std::vector<T> quiet_buffer = {};
-	constexpr auto Silent = Mixer_GetSilentDOSSample<T>();
+	static std::vector<AudioFrame> frames = {};
 
-	if (quiet_buffer.size() < num_samples) {
-		quiet_buffer.resize(num_samples, Silent);
+	const auto is_mono = (frame_type == FrameType::Mono);
+
+	const size_t num_frames = num_samples / (is_mono ? 1 : 2);
```

Comment:
minor: `num_frames_for_frame_type` helper function instead?

Reply:
Added; moved it to a `constexpr`.